src/common/proto_type.cpp

```diff
@@ -122,3 +143,110 @@ Value ProtoType::proto_value() {
 
     return v;
 }
+
+bool operator==(const ProtoType &lhs, const ProtoType &rhs) {
+    if (lhs.proto_type.which() != rhs.proto_type.which()) {
+        return false;
+    }
+
+
+
+
+    switch(lhs.proto_type.which()) {
+          case 0: {
+            // both null values
+            return true;
+            break;
+        }
+        case 1: {
+            const bool lhs_b = boost::get<bool>(lhs.proto_type);
+            const bool rhs_b = boost::get<bool>(rhs.proto_type);
+            if(lhs_b == rhs_b) {
+                return true;
+            }
+            return false;
+            break;
+        }
+        case 2: {
+            const std::string lhs_s = boost::get<std::string>(lhs.proto_type);
+            const std::string rhs_s = boost::get<std::string>(rhs.proto_type);
+             if(lhs_s == rhs_s) {
+                return true;
+            }
+            return false;
+            break;
+        }
+         case 3: {
+            char const* lhs_c = boost::get<char const*>(lhs.proto_type);
+            char const* rhs_c = boost::get<char const*>(rhs.proto_type);
+             if(lhs_c == rhs_c) {
+                return true;
+            }
+            return false;
+            break;
+        }
+        case 4: {
+            const int lhs_i = boost::get<int>(lhs.proto_type);
+            const int rhs_i = boost::get<int>(rhs.proto_type);
+            if(lhs_i == rhs_i) {
+                return true;
+            }
+            return false;
+            break;
+
+        }
+        case 5: {
+            const double lhs_i = boost::get<double>(lhs.proto_type);
+            const double rhs_i = boost::get<double>(rhs.proto_type);
+            if(lhs_i == rhs_i) {
+                return true;
+            }
+            return false;
+            break;
+        }
+        case 6: {
+            std::shared_ptr<std::unordered_map<std::string, std::shared_ptr<ProtoType>>> lhs_map =
+                boost::get<std::shared_ptr<std::unordered_map<std::string, std::shared_ptr<ProtoType>>>>(lhs.proto_type);
+            std::shared_ptr<std::unordered_map<std::string, std::shared_ptr<ProtoType>>> rhs_map =
+                boost::get<std::shared_ptr<std::unordered_map<std::string, std::shared_ptr<ProtoType>>>>(rhs.proto_type);
+            
+            auto it1 = *lhs_map->begin();
+            auto it2 = *rhs_map->begin();
+
+
+            ProtoType lhs_type = *lhs_map->at(it1.first);
```

Comment:
The unordered map equals operator didn't work, not 100% sure why but I'm guessing because of the custom type in the map. However looking at this again pretty sure this would only test the equality for the first element in the map so fixed that, and added a unit test so I could confirm it worked :)